server.go

```diff
@@ -1267,9 +1268,10 @@ func (s *Server) NextProto(key string, nph ServeHandler) {
 	s.nextProtos[key] = nph
 }
 
-func (s *Server) hasNextProto(c net.Conn) (proto string) {
+func (s *Server) hasNextProto(c net.Conn) (proto string, err error) {
 	tlsConn, ok := c.(connTLSer)
 	if ok {
+		err = tlsConn.Handshake()
 		proto = tlsConn.ConnectionState().NegotiatedProtocol
```

Comment:
It's weird to do the `ConnectionState()` call after `Handshake()` already might have returned an error. Maybe rewrite the function to:
```go
func (s *Server) getNextProto(c net.Conn) (string, error) {
	if tlsConn, ok := c.(connTLSer); ok {
		if err := tlsConn.Handshake(); err != nil {
			return "", err
		} else 
			return tlsConn.ConnectionState().NegotiatedProtocol, nil
		}
	} else {
		return "", nil
	}
}
```
I also renamed the function to `getNextProto` as it gets the proto, it doesn't check if it has any as the previous name implied. I would expect a function called `hasNextProto` to return a `bool`.

Reply:
I want to keep the named parameters. See [this article](https://blog.minio.io/golang-internals-part-2-nice-benefits-of-named-return-values-1e95305c8687). So a possible implementation of this function could be:
```go
func (s *Server) getNextProto(c net.Conn) (proto string, err error) {
  tlsConn, ok := c.(connTLSer)
  if ok {
    err = tlsConn.Handshake()
    if err == nil {
      proto = tlsConn.ConnectionState().NegotiatedProtocol
    }
  }
  return
}
```

Reply:
Yes that looks good 👍